templates/company/contracts/CompanyTemplate.sol

```diff
@@ -0,0 +1,94 @@
+        BaseTemplate(_daoFactory, _ens, _miniMeFactory, _aragonID)
+        public
+    {
+        // solium-disable-previous-line no-empty-blocks
+    }
+
+    function newTokenAndInstance(string id, address[] holders, uint256[] stakes) public {
+        require(holders.length == stakes.length, ERROR_INVALID_HOLDERS_STAKES_LEN);
+        newToken();
+        newInstance(id, holders, stakes);
+    }
+
+    function newToken() public returns (MiniMeToken) {
+        MiniMeToken token = _createToken(TOKEN_NAME, TOKEN_SYMBOL);
+        _cacheToken(token, msg.sender);
+        return token;
+    }
+
+    function newInstance(string id, address[] holders, uint256[] stakes) public {
+        require(holders.length == stakes.length, ERROR_INVALID_HOLDERS_STAKES_LEN);
+        MiniMeToken token = _popTokenCache(msg.sender);
+
+        // Create DAO and install apps
+        (Kernel dao, ACL acl) = _createDAO();
+        Agent agent = _installDefaultAgentApp(dao);
+        Finance finance = _installFinanceApp(dao, Vault(agent), FINANCE_PERIOD);
+        TokenManager tokenManager = _installTokenManagerApp(dao, token, TOKEN_TRANSFERABLE, TOKEN_MAX_PER_ACCOUNT);
+        Voting voting = _installVotingApp(dao, token, SUPPORT_REQUIRED, MIN_ACCEPTANCE_QUORUM, VOTE_DURATION);
+
+        // Mint tokens
+        _createPermission(acl, tokenManager, tokenManager.MINT_ROLE());
+        for (uint256 i = 0; i < holders.length; i++) {
+            tokenManager.mint(holders[i], stakes[i]);
+        }
+        _removePermission(acl, tokenManager, tokenManager.MINT_ROLE());
+
+        // Set up permissions
+        _createAgentPermissions(acl, agent, voting, voting);
+        _createVaultPermissions(acl, Vault(agent), finance, voting);
+        _createFinancePermissions(acl, finance, voting, voting);
+        _createTokenManagerPermissions(acl, tokenManager, voting);
+        _createEvmScriptsRegistryPermissions(acl, voting, voting);
+        _createVotingPermissions(acl, voting, tokenManager);
+        _transferPermission(acl, voting, dao, dao.APP_MANAGER_ROLE());
+        _transferPermission(acl, voting, acl, acl.CREATE_PERMISSIONS_ROLE());
+
+        _registerID(id, dao);
+    }
+
+    function _createVotingPermissions(ACL acl, Voting voting, TokenManager tokenManager) internal {
+        acl.createPermission(tokenManager, voting, voting.CREATE_VOTES_ROLE(), voting);
+        acl.createPermission(voting, voting, voting.MODIFY_QUORUM_ROLE(), voting);
+        acl.createPermission(voting, voting, voting.MODIFY_SUPPORT_ROLE(), voting);
+    }
+
+    function _createTokenManagerPermissions(ACL acl, TokenManager tokenManager, Voting voting) internal {
+        acl.createPermission(voting, tokenManager, tokenManager.BURN_ROLE(), voting);
+        acl.createPermission(voting, tokenManager, tokenManager.MINT_ROLE(), voting);
+    }
+
+    function _cacheToken(MiniMeToken token, address owner) internal {
```

Comment:
I wonder if these 2 functions could be moved to `BaseTemplate` or to a library for them to be reused. Seems like a common use case.

Reply:
All the caches work differently in the different templates, I don't think it could be reusable

Reply:
Well, I would say these 4 are quite similar :wink: : 
https://github.com/aragon/dao-templates/blob/next/templates/company/contracts/CompanyTemplate.sol#L83
https://github.com/aragon/dao-templates/blob/next/templates/democracy/contracts/DemocracyTemplate.sol#L79
https://github.com/aragon/dao-templates/blob/next/templates/membership/contracts/MembershipTemplate.sol#L87
https://github.com/aragon/dao-templates/blob/next/templates/multisig/contracts/MultisigTemplate.sol#L76
https://github.com/aragon/dao-templates/blob/next/templates/reputation/contracts/ReputationTemplate.sol

And I guess there will eventually be more like these ones.